### PR TITLE
Set the mapped security roles of the user so these can be used by the…

### LINKED ISSUE #1568: [BUG] Unexpected role set used for document level security.
**Describe the bug:**
Opensearch Document Level Security (DLS) leverages the backend roles of a user to facilitate authorisation decisions. I was expecting the Opensearch roles to be utilised for this purpose; similar to managing the index level permissions. I'm looking at the parameter substitution as outlined here: [https://opensearch.org/docs/latest/security-plugin/access-control/document-level-security/#parameter-substitution](https://opensearch.org/docs/latest/security-plugin/access-control/document-level-security/#parameter-substitution).

I've got an issue currently where the backend roles are inflexibly named and require the role mapping functionality to align to something more useful within the DLS. I've done a bit of a hunt around and it looks like it comes down to the following bits of code: [https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java#L924-L931](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java#L924-L931) and [https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/user/User.java#L66-L70](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/user/User.java#L66-L70).

I understand that there will be a fair amount of existing dependencies leveraging this functionality in this state. I'm proposing the exposure of the Opensearch specific roles as an additional substitution value to remedy; let me know if there's something I'm missing on this one.

**Expected behaviour:**
The `${user.roles}` having resolved to the set of roles which is the output of the role mapping configuration within Opensearch. 

**Plugins:**
Default docker image 

**Host/Environment (please complete the following information):**
OpenSearch v.1.2.3 - Docker image

**Additional context:**
For additional context, I've got a dependency on a backend authentication system which it is impractical to manipulate to align to what is required to facilitate the DLS. The role mapping functional allows for the translation of the backend roles over to something more practical but doesn't seem to be utilised for the DLS authorisation decisions.

### PR DESCRIPTION
**Pull request by ch-govau**
… DLS privileges evaluator. Allow security roles to be used for DLS parameter substitution. Fixes opensearch-project/security/#1568

Signed-off-by: Caitlin Harper <caitlin.harper@defence.gov.au>

### Description
* Category (Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation)
Bug Fix /Enhancement

* Why these changes are required?
The changes are required so that we can map the backend roles to more appropriately named roles via the roles.mapping and these security roles can be used in the DLS privileges evaluation. 

* What is the old behavior before changes and new behavior after changes?
Old behaviour:
security roles were not added to the user object prior to the DLS privileges evaluation

New behaviour:
mapped roles are added to the user object for DLS  privileges evaluation
a new option for parameter substitution in the DLS query has been added. This is ${user.securityRoles} and is the mapped roles.

Note: there is no change to the existing ${user.roles} substitution. It remains the same.

### Issues Resolved
* Resolves #1568 

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
Manual test steps:
Testing was completed for both basic auth and OIDC/jwt auth. Note that both users had read access to the index.
1. Push 2 documents into the index, one with DLS required role = ROLE1 and the other with ROLE2.
example: 
```
curl https://localhost:9200/myindex/_doc?pretty -u admin:$PASS -XPOST -H "Content-Type: application/json" -d @doc1.json
```
doc1.json
```json
{
	"security": {
                        "exclusive": [
                            "ROLE1",
                        ],
                        "num_exclusive": 1
          },
	  "info":{
		"generalinfo": "this document should only be available to those with ROLE1"
	  }
}
```
2. Add the roles ROLE1, ROLE2 and the read to the roles.yml
example:
```yaml
ROLE1:
  reserved: false
  hidden: false
  cluster_permissions: []
  index_permissions: []
  tenant_permissions: []
  static: false

ROLE2:
  reserved: false
  hidden: false
  cluster_permissions: []
  index_permissions: []
  tenant_permissions: []
  static: false

myindex_read:
  reserved: false
  hidden: false
  cluster_permissions:
  - "indices:data/read/scroll/clear"
  - "indices:data/read/msearch"
  - "indices:data/write/bulk"
  - "indices:data/read/scroll"
  - "indices:data/read/mget"
  index_permissions:
  - index_patterns:
    - "myindex"
    dls: "{ \"bool\": { \"filter\": [\n   {\"terms_set\": {\"security.exclusive.keyword\": {\n\
      \ \"terms\": [${user.securityRoles}],\n \"minimum_should_match_field\": \"security.num_exclusive\"\
      \n }}}\n ]}}\n"
    fls: []
    masked_fields: []
    allowed_actions:
    - "read"
```
3. Add the role mappings to the backend roles in roles_mapping.yml
```yaml
ROLE1:
  hosts: []
  users: []
  reserved: false
  hidden: false
  backend_roles:
  - "client-id/role-1"
  and_backend_roles: []

ROLE2:
  hosts: []
  users: []
  reserved: false
  hidden: false
  backend_roles:
  - "client-id/role-2"
  and_backend_roles: []
```
4. Also add the following user for the internal/basic auth testing
```yaml
user1:
  hash: "<password-hash>"
  reserved: false
  backend_roles:
    - "client-id/role-1"
    - "myindex_read"
  description: "user1"
```
5. Note that we also have the test server set up to use keycloak and user2 has the following backend roles: "myindex_read", "client-id/role-1", "client-id/role-2".
6. Run the security config reload script
```
./plugins/opensearch-security/tools/securityadmin.sh" -cd "./plugins/opensearch-security/securityconfig" -icl -key "./config/kirk-key.pem" -cert "./config/kirk.pem" -cacert "./config/root-ca.pem" -nhnv
```
7. Test basic auth for user 1
```
curl https://localhost:9200/myindex/_search?pretty --insecure -u user1:user1
```
Expected result: This should return 1 document, doc1
8. Test jwt for user2
```
curl -H "Authorization: Bearer $TOKEN" https://localhost:9200/myindex/_search?pretty --insecure
```
Expected result: This should return 2 documents.
9. Also tested with another user that has neither ROLE1 or ROLE2 and confirm they have access to zero documents.

### Documentation
This will require an update to the DLS parameter substitution documentation to include the new ${user.securityRoles}.
The update to the documentation has been submitted here: https://github.com/opensearch-project/documentation-website/pull/420

### Check List
- [x] New functionality includes testing
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ae-govau** [comment]
Gentle ping - our stakeholders are still keen on this. Would appreciate feedback (positive or negative) on the approach and whether anything additional is needed.

I note the failing "CodeQL" check - but I note that what it is complaining about is not relevant to (ie is not changed by) this PR.

**Comment by cliu123** [comment]
Thanks for contributing! Can you add UTs for all possible use cases?

**Comment by ch-govau** [comment]
> Thanks for contributing! Can you add UTs for all possible use cases?

Can do.

**Comment by ch-govau** [comment]
Tests added, will rebase and squash when everyone is happy.

**Comment by ch-govau** [comment]
Any thoughts/comments stopping this from being merged. I'm happy to make changes if required? Also the documentation update has been completed by my colleague and is linked to the PR.

**Comment by peternied** [comment]
@opensearch-project/security can we review this PR?

**Comment by ae-govau** [comment]
Hey all, this has been sitting stagnant for a while. @ch-govau has added the tests, and @mm-govau has added the documentation in https://github.com/opensearch-project/documentation-website/pull/420 which is already approved.

I think all we are waiting on is an approval here?  Any attention for review for would be much appreciated. Outside of the tests, it's 10 lines of new code, so hoping it's not a hard review task.

Thanks again.

**Comment by peternied** [comment]
@cliu123 can you review this?

File: src/test/resources/dlsfls/roles.yml
Comment (cliu123) [inline src/test/resources/dlsfls/roles.yml:1389]: Please remove unnecessary indentation.

File: src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java
Comment (cliu123) [inline src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java:936]: What is `user_securityRoles`

File: src/test/java/org/opensearch/security/dlic/dlsfls/DlsPropsReplaceTest.java
Comment (cliu123) [inline src/test/java/org/opensearch/security/dlic/dlsfls/DlsPropsReplaceTest.java:67]: How does this ensure which doc is hit? Would it be more appropriate to assert `amount`?

**Comment by cliu123** [comment]
Could you please sign all the commits?

File: src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java
Comment (ch-govau) [reply src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java]: I included the `user_securityRoles` as the underscore seemed to be supported for the other two parameter substitutions for `user_name` and `user_roles`. I have removed now.

File: src/test/java/org/opensearch/security/dlic/dlsfls/DlsPropsReplaceTest.java
Comment (ch-govau) [reply src/test/java/org/opensearch/security/dlic/dlsfls/DlsPropsReplaceTest.java:67]: updated to assert the amount.

**Comment by ch-govau** [comment]
> Could you please sign all the commits?

sorry about that. have rebased and the commit is signed. Thanks so much for reviewing this. It is very much appreciated :)

File: src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java
Comment (cliu123) [reply src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java]: I didn't mean that `user_securityRoles` shouldn't be removed. I actually have a question on this. Is it needed here? Can we make sure that it exists in DLS query and index pattern, and it should be supported?

Comment (ch-govau) [reply src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java]: I can add it back in and would you like me to add a test for it or is there something else I also need to add?

**Review by peternied — DISMISSED** [review]

File: src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java
Comment (cliu123) [reply src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java]: Sorry for the confusion if the question wasn't clear! I see the following in the tests. `securityRoles` are being populated in the requests. Is `user_securityRoles` another field supported in requests? Or is it an alternative of `securityRoles` in requests? Could you please help me understand this?
```
       tc.index(new IndexRequest("prop-mapped").type("_doc").setRefreshPolicy(RefreshPolicy.IMMEDIATE)
            .source("{\"securityRole\": \"opendistro_security_mapped\", \"amount\": 6060}", XContentType.JSON)).actionGet();
        tc.index(new IndexRequest("prop-mapped").type("_doc").setRefreshPolicy(RefreshPolicy.IMMEDIATE)
            .source("{\"securityRole\": \"not_assigned\", \"amount\": 7070}", XContentType.JSON)).actionGet();
```

Comment (ch-govau) [reply src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java]: I'm only testing the `user.securityRoles`in my test with the following DLS query
```
    dls: "{\"terms\" : { \"securityRole\" : [${user.securityRoles}]}}"
```
I then add two files to my `prop-mapped` index, each requiring a different security role. I assigned the backend role of `prolemapped` to the user and this maps to the security role `opendistro_security_mapped`. So the user is able to read only the one document that they have the role for. If I wanted to use the `user_securityRoles` substitution then I could modify the DLS query to the following:
```
dls: "{\"terms\" : { \"securityRole\" : [${user_securityRoles}]}}"
```
If I add the underscore substitution option back in then this would have the same outcome for the user.

My guess is that the underscore substitution was to maintain backwards compatibility with older versions as it is an undocumented feature.

So, I can add it back in if we want to support both options?

Comment (cliu123) [reply src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java]: Ok. Please add it back then. Thanks!

Comment (ch-govau) [reply src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java:932]: Have added it back in. Thanks @cliu123 and @peternied for looking at this for us :)

**Review by peternied — APPROVED** [review]

**Comment by ae-govau** [comment]
@cliu123 - is anything else needed on this PR?  I think @ch-govau has addressed all comments and this can be merged after a 2nd approval. Thanks for your help.

**Comment by cliu123** [comment]
> @cliu123 - is anything else needed on this PR? I think @ch-govau has addressed all comments and this can be merged after a 2nd approval. Thanks for your help.

LGTM! Thanks for the contribution!

**Review by cliu123 — APPROVED** [review]

**Pull request merged**
